chapter/related-work.md

```diff
@@ -36,10 +36,94 @@ Additionally, this part aims to recognize alternative approaches to the LSP, in
 
 ## Alternative approaches
 
-### Platform plugins
+### LSP Extensions
 
-### Legacy protocols
+The LSP defines a large range of commands and capabilities which is continuously being extended by the maintainers of the protocol.
+Yet, occasionally server developers find themselves in need of functionality not yet present in the protocol.
+For example the LSP does not provide commands to exchange binary data such as files.
+In [@sec:language-independence] the CodeCompass Language Server was introduced.
+A stern feature of this server is the ability to generate and show diagrams in SVG format.
+However, the LSP does not define the concept of *requesting diagrams*.
+In particular Mészáros et al. describe different shortcomings of the LSP :
 
-### LSP Extensions
+1. > "LSP doesn’t have a feature to place a context menu at an arbitrary spot in the document"
+   Context menu entries are implemented by clients based on the agreed upon capabilities of the server.
+   Undefined capabilities cannot be added to the context menu.
+
+   In the case of CodeCompass the developers made up for this by using the code completion feature as an alternative dynamic context menu.
+
+2. "LSP does not support displaying pictures (diagrams)".
+   CodeCompass generates diagrams for selected code.
+   Yet, there is no image transfer included with the LSP.
+   Since the LSP is based on JSON-RPC messages, the authors' solution was to define a new command, specifically designed to tackle this non-standard use case.
```

Comment:
```suggestion
1. > "LSP doesn’t have a feature to place a context menu at an arbitrary spot in the document"

   Context menu entries are implemented by clients based on the agreed upon capabilities of the server.
   Undefined capabilities cannot be added to the context menu.

   In the case of CodeCompass the developers made up for this by using the code completion feature as an alternative dynamic context menu.

2. > "LSP does not support displaying pictures (diagrams)".

   CodeCompass generates diagrams for selected code.
   Yet, there is no image transfer included with the LSP.
   Since the LSP is based on JSON-RPC messages, the authors' solution was to define a new command, specifically designed to tackle this non-standard use case.
```